aten/src/ATen/native/ReduceOps.cpp

```diff
@@ -1434,23 +1443,16 @@ Tensor amin(const Tensor& self, IntArrayRef dim, bool keepdim) {
   return at::amin_out(result, self, dim, keepdim);
 }
 
-Tensor &amax_out(const Tensor& self, IntArrayRef dim, bool keepdim, Tensor& result) {
+TORCH_IMPL_FUNC(amax_out) (const Tensor& self, IntArrayRef dim, bool keepdim, const Tensor& result) {
   TORCH_CHECK(self.scalar_type() == result.scalar_type(), "Expected the dtype for input and out to match, but got ",
```

Comment:
nit: this is a dtype check, so it should go in the meta function.

you pointed out before that "result" isn't an input to the meta function, since the meta function needs to used in both the out-of-place and the out= case.

The way you can differentiate is by calling `maybe_get_output()[0]` inside the meta function - if it's a valid tensor (`maybe_get_output()[0].defined()`), then you've been passed a preallocated out= tensor and you can run the check.